File: same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
if roots have same value, and likewise for p.left q.left, and likewise for p.riht q.right. then they're the same
'''
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False
        assert p and q
        issame = p.val == q.val
        issame = issame and self.isSameTree(p.left,q.left)
        issame = issame and self.isSameTree(p.right,q.right)
        return issame

File: lc/same-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// s(p, q)
// boundary conditions: 
//     one may be nil, and the other not. 
//     both nil is ok
// base case returns false when p and q are not the same
// recursive steps is simple: check left and right
    //     1           1
    //   /   \
    //  2     3       2     
func isSameTree(p *TreeNode, q *TreeNode) bool {
    if p == nil && q == nil { return true }
    if p == nil || q == nil { return false }
    if p.Val != q.Val { return false }
    return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}